plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Prism/Mirror.hs

```diff
@@ -0,0 +1,92 @@
+{-# LANGUAGE DataKinds          #-}
+{-# LANGUAGE DeriveAnyClass     #-}
+{-# LANGUAGE DeriveGeneric      #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts   #-}
+{-# LANGUAGE MonoLocalBinds     #-}
+{-# LANGUAGE TemplateHaskell    #-}
+{-# LANGUAGE TypeApplications   #-}
+{-# LANGUAGE TypeFamilies       #-}
+{-# LANGUAGE TypeOperators      #-}
+-- | The Atala mirror application that initialises the state machine
+module Language.PlutusTx.Coordination.Contracts.Prism.Mirror(
+    MirrorSchema
+    , MirrorError(..)
+    , mirror
+    ) where
+
+import           Control.Monad                                               (void)
+import           Data.Aeson                                                  (FromJSON, ToJSON)
+import           GHC.Generics                                                (Generic)
+import           Language.Plutus.Contract
+import           Language.Plutus.Contract.StateMachine                       (SMContractError, runInitialise, runStep)
+import           Language.PlutusTx.Coordination.Contracts.Prism.Credential   (Credential (..), CredentialAuthority (..))
+import qualified Language.PlutusTx.Coordination.Contracts.Prism.Credential   as Credential
+import           Language.PlutusTx.Coordination.Contracts.Prism.StateMachine as StateMachine
+import           Ledger                                                      (txId)
+import qualified Ledger.Constraints                                          as Constraints
+import           Ledger.Crypto                                               (PubKeyHash, pubKeyHash)
+import qualified Ledger.Typed.Scripts                                        as Scripts
+import           Ledger.Value                                                (TokenName)
+
+type MirrorSchema =
+    BlockchainActions
+        .\/ Endpoint "issue" (TokenName, PubKeyHash) -- lock a single token in a state machine tied to the client address
+        .\/ Endpoint "revoke" (TokenName, PubKeyHash) -- revoke a token from a state machine
+
+mirror ::
+    ( HasBlockchainActions s
+    , HasEndpoint "revoke" (TokenName, PubKeyHash) s
+    , HasEndpoint "issue" (TokenName, PubKeyHash) s
+    )
+    => Contract s MirrorError ()
+mirror = do
+    authority <- mapError SetupError $ CredentialAuthority . pubKeyHash <$> ownPubKey
+    let go = (createTokens authority `select` revokeToken authority) >> go
```

Comment:
`forever` didn't use to work in the "old" implementation of `Contract`, before I switched over to `freer-simple` effects (because it uses `Applicative`, not `Monad`, resulting in a loop). It should be fine now though, I'll check.

Reply:
Yep it works